### ISSUE TITLE
Transformer.Train per-call heap retention still leaks 79 KB/call on AiDotNet 0.198.0 + Tensors 0.81.0 (down from 1.55 MB on 0.75.4 but not eliminated)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1340]
## Summary

Follow-up to #1227 (Closed 2026-05-11). The L=4 Transformer.Train leak was 1.55 MB/call on Tensors 0.75.4. **On the current pinned versions (AiDotNet 0.198.0 / Tensors 0.81.0), the leak is reduced ~20× but still measurable at ~79 KB/call.**

Symptom: HarmonicEngine PAPER-A Path B Pareto sweep OOMs `testhost` at peak working-set ~7.5 GB on a 164k-param Transformer training 5,760 samples × 2 epochs across 4 parallel seeds. HRE retrieval LM (zero-backprop) on the same task peaks at 40 MB total — a **~150× memory ratio**.

## Measurement environment

- AiDotNet **0.198.0**, AiDotNet.Tensors **0.81.0**
- Windows 11, .NET 9 (HarmonicEngine pinned)
- 16 GB host
- Standalone xUnit test, no GPU forced (OpenCL auto-detected but training stays on managed CPU path)

## Repro (Verbatim from the in-repo measurement test)

Test file: `tests/Learning/Phase_AiDotNet_TransformerTrain_LeakMeasurement.cs` (committed in HarmonicEngine).

The chain is identical to the PAPER-A Path B consumer config:

```csharp
var chain = new List<ILayer<float>>
{
    new InputLayer<float>(64),
    new EmbeddingLayer<float>(256, 128),
    new PositionalEncodingLayer<float>(64, 128),
    // 2x:
    new MultiHeadAttentionLayer<float>(2, 64, (IActivationFunction<float>)new IdentityActivation<float>()),
    new LayerNormalizationLayer<float>(),
    new DenseLayer<float>(256, new ReLUActivation<float>()),
    new DenseLayer<float>(128, new IdentityActivation<float>()),
    new LayerNormalizationLayer<float>(),
    // x 2
    new SequenceTokenSliceLayer<float>(SequenceTokenSliceLayer<float>.Position.Last),
    new DenseLayer<float>(256, new IdentityActivation<float>()),
    new ActivationLayer<float>(new SoftmaxActivation<float>()),
};

var arch = new NeuralNetworkArchitecture<float>(
    inputType: InputType.OneDimensional,
    taskType: NeuralNetworkTaskType.MultiClassClassification,
    inputSize: 64, outputSize: 256, layers: chain);

var optimizer = new AdamOptimizer<float, Tensor<float>, Tensor<float>>(null,
    new AdamOptimizerOptions<float, Tensor<float>, Tensor<float>>
    { InitialLearningRate = 1e-3, MaxIterations = 1, UseAdaptiveLearningRate = false });
var network = new FeedForwardNeuralNetwork<float>(arch, optimizer,
    lossFunction: new CategoricalCrossEntropyLoss<float>());
```

Then 500 `network.Train(input, target)` calls with the same fixed `Tensor<float>([1, 64])` input and one-hot target. Forced `GC.Collect(2, Aggressive)` + `WaitForPendingFinalizers` + `GetTotalMemory(forceFullCollection: true)` sampled every 50 calls.

Output (verbatim):

```
Trainable parameters: 164,096
Model weight size: 0.63 MB
Baseline: WS=202.5 MB ; ManagedHeap=14.3 MB ; AllocatedTotal=89.0 MB

call    WS_MB   Heap_MB   Alloc_MB  Gen2  per-call_KB
50      268     53        112       11    358
100     267     52        112        9    -24
150     272     54        112        7     66
200     275     56        112        6     53
250     279     58        112        4     57
300     282     61        112        4     57
350     285     63        112        4     54
400     288     65        112        4     53
450     291     68        112        4     58
500     297     72        112        4     56

=== SUMMARY ===
Total Train calls:           500
Wall:                        16.3 s (32.6 ms/call)
Working-set delta:           66.0 MB
Managed-heap delta:          38.7 MB
Total allocated:             6046.5 MB
Avg per-call heap retention: 78.8 KB
Projected at 50k calls:      3.8 GB
Projected at 200k calls:     15.0 GB
```

## Key facts

1. **Per-call retention is linear after warm-up.** Calls 200-500 retain a tight 53-58 KB/call (samples 4-9 average to ~55 KB/call). The first sample (358 KB/call) is warm-up; one sample even shows -24 KB (a Gen2 reclaim cycle). Steady-state is the relevant figure for production.

2. **Gen2 collection frequency drops over time.** 11 Gen2s in calls 1-50, then drops to 4 every 50 calls. The retained tensors are surviving Gen2 promotion.

3. **Total allocated stays flat per-window.** Each 50-call window allocates ~112 MB (~2.2 MB/call). But only ~56 KB/call survives. The allocation pressure is fine; the retention is the bug.

4. **Working-set tracks managed-heap.** Same pattern as #1227 — this is a managed-heap leak, not a native leak.

## Why this is still a P1 for AiDotNet consumers

For training jobs of any meaningful size:

| Training scale | Train calls | Projected heap |
|----------------|-------------|----------------|
| 100KB byte LM, 2 epochs, stride=8 | ~22,500 | **~1.7 GB** |
| 1MB byte LM, 2 epochs, stride=8 | ~225,000 | **~17 GB** (OOM on 16 GB host) |
| 10MB byte LM, 2 epochs, stride=8 | ~2,250,000 | **~170 GB** (host crash) |

For HarmonicEngine specifically: the PAPER-A Path B Pareto sweep OOM'd `testhost` at ~7.5 GB peak with 4 parallel seeds × 5760 calls/seed = 23,040 calls. **The math matches** (23,040 × 79 KB × 4 parallel = ~7.3 GB).

The ratio comparison HRE retrieval LM (the head-to-head against this Transformer) peaks at 40 MB total memory on the same byte-LM task. ~150× ratio for an equivalent inference quality is the actual cost of the leak in published numbers.

## Suggested next-step

Per the #1227 dotnet-trace from 2026-05-01, the leak was localized to `GradientTape<T>` static fields:
```
_cachedArena, _cachedDelegateChain, _entries
```

The `Clone()`-per-Train pattern was profiled at ~15× per call. The fix attempted in PR #1232 / Tensors 0.69.1 addressed L=1 only. The 0.80.x line dropped 1.55 MB/call → 79 KB/call (~20× improvement, presumably from #283 / #284 work) but the floor isn't 0 yet.

Likely remaining culprit: per-Backward saved-for-backward intermediates that aren't released after `ComputeGradientsViaGraph` returns. Each Transformer encoder block has ~15 of these (Q/K/V/scores/attn-out/FFN-hidden/FFN-out/residuals/norms); at 2 layers × 15 = ~30 tensor objects retained per call. At ~2 KB each (small `Tensor<float>` object overhead + a few elements survived) that gives ~60 KB — matches the observed 56 KB/call steady-state.

Recommended fix surface: `NeuralNetworkBase<T>.TrainWithTape` should explicitly clear `tape.SavedInputs`/`tape.SavedOutputs`/the per-step closures after `ComputeGradientsViaGraph` returns and before scope exit. If `GradientTape<T>` already implements `IDisposable`, `using var tape = ...` plus an explicit `tape.SavedTensors?.Clear()` in `Dispose` should bring the rate to 0.

## Workarounds being shipped downstream (HarmonicEngine)

Until this lands, HarmonicEngine ships:

1. Per-cell process isolation (one testhost subprocess per seed) so the leak cap is bounded to one seed's worth (~1.7 GB for 100KB cells, still fits 16 GB).
2. Explicit `GC.Collect(2, Aggressive)` every 500 Train calls inside `AiDotNetNetworkLMPredictor.Build` to force Gen2 promotion regularly and keep working-set under control.
3. The PAPER-A Path B Pareto quick-cell is downsized to 50KB × 2 seeds × serial.

These are reproducible band-aids; the underlying leak still rules out 1MB+ byte-LM head-to-head Transformer training within the host's commit limit.

## Repro test in HarmonicEngine

`tests/Learning/Phase_AiDotNet_TransformerTrain_LeakMeasurement.cs` — committable to AiDotNet test suite if useful for regression coverage. Output above is reproducible end-to-end on a clean checkout.

**Comment by ooples** [comment]
## Fix landed in [AiDotNet.Tensors#360](https://github.com/ooples/AiDotNet.Tensors/pull/360)

Verified locally:
- **Before fix:** 78.8 KB/call (this issue's measurement)
- **After fix:** 0-1 B/call (2 independent runs, 500 iters each)

### What was fixed
`CompiledDelegateChain<T>._steps[]` held 4 strong refs per step (Output / Inputs / Backward / SavedState). On `GradientTape.Dispose` the chain *pointer* was nulled but the *array* still pinned ~30 tensor references via the BackwardStep struct fields. Any reference path that retained the chain instance (finalizer ordering, async continuations, BackwardScratch pool entries) kept the forward intermediates alive past Gen2 GC.

The fix adds `CompiledDelegateChain<T>.Clear()` that zeros each `_steps[i]` (sets to `default(BackwardStep<T>)`) and calls it from `GradientTape<T>.Dispose()` before nulling the chain pointer. Safe because Dispose is on the discard path — never called during persistent-tape replay.

### Regression test
`GradientTapeLeakTests.TrainStep_Transformer_L2_PersistentTape_Issue1340_RetentionUnder20KbPerCall` — 2-encoder-block chain matching this issue's consumer config (164k params), 500 Train iterations, asserts <20 KB/call retention (pre-fix ~79 KB → fail, post-fix 0-1 B → pass).

### Downstream
No AiDotNet repo code change needed — the existing `using var tape = new GradientTape<T>()` in `NeuralNetworkBase.TrainWithTape` already invokes Dispose. Once Tensors#360 merges and releases as 0.82.0, AiDotNet just needs to bump the `<PackageVersion>` reference.

### HarmonicEngine impact
Once the package bump lands, the PAPER-A Path B Pareto sweep (currently blocked at 50KB × 2 seeds × serial due to ~7.5 GB peak) should run 8 seeds × 1MB cleanly within the 16 GB host commit limit.

**Comment by ooples** [comment]
## Likely root cause + fix design

Per memory notes from prior session investigation: the Tensors PR #360 fixed the **`CompiledDelegateChain._steps[]` retention** (FFN-side), bringing the leak from ~1.55 MB/call (0.75.4) → ~79 KB/call (0.198.0/0.81.0).

The remaining 79 KB/call is consistent with **MultiHeadAttentionLayer cache field retention**:

`src/NeuralNetworks/Layers/MultiHeadAttentionLayer.cs` stores per-call state in:
- `_lastQueryInput`, `_lastKeyInput` — input snapshots
- `_lastProjectedQueries`, `_lastProjectedKeys`, `_lastProjectedValues` — [batch, heads, seq, head_dim] each
- `_lastAttentionScores` — [batch, heads, seq, seq]
- `_lastHeadOutputs` — list of stacked-heads tensors
- `_lastAttentionContext` — [batch, seq, embeddingDim]

These are intentional (used by Backward pass), but they're never CLEARED. Each Train call replaces them with new tensor references, but **the previous Train's tensors are retained until the NEXT Train call writes over them**. Worse — they're never cleared even after `SetTrainingMode(false)` is called.

For a 2-layer Transformer with dModel=128 / ctxLen=64 / heads=2 / vocab=256:
- 5 MHA cache tensors × 2 layers × 4 bytes per float ≈ several KB to tens of KB per layer per call
- At ~80 KB/call across 2 layers, ~40 KB per MHA layer per call — matches the measured 79 KB/call

This intersects with **#1359** (MHA cache lifetime bugs) where the agent found stale-reference NaN for deeper-layer MHA and zero-readback for head > 0. Both bugs share the same root: the cache fields lack disciplined lifecycle management.

## Recommended fix

Add explicit clear-on-eval to `MultiHeadAttentionLayer<T>.SetTrainingMode(bool)`:

```csharp
public override void SetTrainingMode(bool training)
{
    base.SetTrainingMode(training);
    if (!training)
    {
        // Drop all training-time caches so eval mode doesn't retain
        // forward-pass intermediates. Prevents per-call leak (#1340) and
        // stale-reference bugs (#1359).
        _lastQueryInput = null;
        _lastKeyInput = null;
        _lastProjectedQueries = null;
        _lastProjectedKeys = null;
        _lastProjectedValues = null;
        _lastAttentionScores = null;
        _lastHeadOutputs = null;
        _lastAttentionContext = null;
        // … and any other _last* fields …
    }
}
```

Additionally, for training mode where caches MUST be retained for backward:
- Add `ClearTrainingCaches()` public method that consumers can call between training steps
- Document explicitly that caches grow per-call and consumers must call this to bound working set

## Verification

Add `MultiHeadAttentionLayer_Eval_DoesNotRetainTrainingCaches` regression test:

```csharp
[Fact]
public void EvalMode_ClearsTrainingCaches()
{
    var mha = new MultiHeadAttentionLayer<float>(ctxLen: 64, embDim: 128, heads: 2);
    mha.SetTrainingMode(true);
    mha.Forward(input);
    Assert.NotNull(mha._lastAttentionScores);  // cache populated
    
    mha.SetTrainingMode(false);
    Assert.Null(mha._lastAttentionScores);     // cache cleared
    Assert.Null(mha._lastQueryInput);
    Assert.Null(mha._lastProjectedQueries);
    // ... all _last* fields ...
}
```

Then re-run the HE per-call leak measurement test (`Phase_AiDotNet_TransformerTrain_LeakMeasurement` if it still exists, or equivalent). Expect 79 KB/call → near 0 if MHA caches are the only remaining source.

## Related

- #1359 (MHA cache state corruption — same root: undisciplined cache lifecycle)
- Tensors PR #360 (closed earlier FFN-side leak)
- AiDotNet#1227 (predecessor)

🤖 follow-up analysis by Claude Code
